File: solutions/23.cpp
#include <bits/stdc++.h>
using namespace std

//Method 1

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* a, ListNode* b) {
        if(a == nullptr)
            return b;
        if(b == nullptr)
            return a;
        ListNode* c;
        if(a->val < b->val){
            c = a;
            c->next = mergeTwoLists(a->next, b);
        }
        else{
            c = b;
            c->next = mergeTwoLists(a,b->next);
        }
        return c;
    }
};

File: solutions/23.java
import java.util.*;

class Solution {
  public ListNode rotateRight(ListNode head, int k) {
    if (head == null) return null;

    int len = 1;
    ListNode fast = head, slow = head;
    while (fast.next != null) {
      len++;
      fast = fast.next;
    }

    for (int i = 1; i < len - k % len; i++) slow = slow.next;

    fast.next = head;
    head = slow.next;
    slow.next = null;

    return head;
  }
}
